Add `debug-refcell` feature to libcore

Aaron1011 opened the pull request:
See https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Attaching.20backtraces.20to.20RefCell/near/226273614
for some background discussion

This PR adds a new off-by-default feature `debug-refcell` to libcore.
When enabled, this feature stores additional debugging information in
`RefCell`. This information is included in the panic message when
`borrow()` or `borrow_mut()` panics, to make it easier to track down the
source of the issue.

Currently, we store the caller location for the earliest active borrow.
This has a number of advantages:
* There is only a constant amount of overhead per `RefCell`
* We don't need any heap memory, so it can easily be implemented in core
* Since we are storing the *earliest* active borrow, we don't need any
  extra logic in the `Drop` implementation for `Ref` and `RefMut`

Limitations:
* We only store the caller location, not a full `Backtrace`. Until
  we get support for `Backtrace` in libcore, this is the best tha we can
do.
* The captured location is only displayed when `borrow()` or
  `borrow_mut()` panics. If a crate calls `try_borrow().unwrap()`
  or `try_borrow_mut().unwrap()`, this extra information will be lost.

To make testing easier, I've enabled the `debug-refcell` feature by
default. I'm not sure how to write a test for this feature - we would
need to rebuild core from the test framework, and create a separate
sysroot.

Since this feature will be off-by-default, users will need to use
`xargo` or `cargo -Z build-std` to enable this feature. For users using
a prebuilt standard library, this feature will be disabled with zero
overhead.

I've created a simple test program:

```rust
use std::cell::RefCell;

fn main() {
    let _ = std::panic::catch_unwind(|| {
        let val = RefCell::new(true);
        let _first = val.borrow();
        let _second = val.borrow();
        let _third = val.borrow_mut();
    });

    let _ = std::panic::catch_unwind(|| {
        let val  = RefCell::new(true);
        let first = val.borrow_mut();
        drop(first);

        let _second = val.borrow_mut();

        let _thid = val.borrow();
    });
}
```

which produces the following output:

```
thread 'main' panicked at 'already borrowed: BorrowMutError at refcell_test.rs:6:26', refcell_test.rs:8:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'already mutably borrowed: BorrowError at refcell_test.rs:16:27', refcell_test.rs:18:25
```

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1930143797) [(plain)](https://github.com/rust-lang/rust/commit/bbbde1019ec6938b2f1e952b1dade8f00ad0e47c/checks/1930143797/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1930271117) [(plain)](https://github.com/rust-lang/rust/commit/1c032f64a6c07a682514b70512daac37479441c2/checks/1930271117/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/core/src/cell.rs at line 666:
         RefCell {
             value: UnsafeCell::new(value),
             borrow: Cell::new(UNUSED),
-            #[cfg(feature = "debug_refcell")] first_caller: Cell::new(None),
+            #[cfg(feature = "debug_refcell")]
+            first_caller: Cell::new(None),
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/core/src/cell.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
     }
 
Diff in /checkout/library/core/src/cell.rs at line 806:
Diff in /checkout/library/core/src/cell.rs at line 806:
         #[cfg(feature = "debug_refcell")]
         match self.try_borrow() {
             Ok(b) => return b,
-            Err(e) => panic!("already mutably borrowed: {:?} at {}", e, self.first_caller.get().unwrap())
+            Err(e) => {
+                panic!("already mutably borrowed: {:?} at {}", e, self.first_caller.get().unwrap())
         }
 
 
         #[cfg(not(feature = "debug_refcell"))]
Diff in /checkout/library/core/src/cell.rs at line 900:
         #[cfg(feature = "debug_refcell")]
         match self.try_borrow_mut() {
             Ok(b) => return b,
-            Err(e) => panic!("already borrowed: {:?} at {}", e, self.first_caller.get().unwrap())
+            Err(e) => panic!("already borrowed: {:?} at {}", e, self.first_caller.get().unwrap()),
 
 
         #[cfg(not(feature = "debug_refcell"))]
Build completed unsuccessfully in 0:00:16
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1930372164) [(plain)](https://github.com/rust-lang/rust/commit/dc2fa13bc7d49d9d2e98843e60c8889d0648d428/checks/1930372164/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 7300/11462
.................................................................................................... 7400/11462
.................................................................................................... 7500/11462
..........................................................................................i....ii... 7600/11462
..........F.FF..F.........................................ii........................................ 7700/11462
.............................................................i...................................... 7900/11462
......................................................i............................................. 8000/11462
.................................................................................................... 8100/11462
.................................................................................................... 8200/11462
---
.................................................................................................... 9200/11462
.................................................................................................... 9300/11462
.................................................................................................... 9400/11462
....................i......i........................................................................ 9500/11462
...........................................................iiiiiii..iiiiii.i........................ 9600/11462
.................................................................................................... 9800/11462
.................................................................................................... 9900/11462
.................................................................................................... 10000/11462
.................................................................................................... 10100/11462
---

---- [ui] ui/not-panic/not-panic-safe-2.rs stdout ----
diff of stderr:

25    = note: required because it appears within the type `RefCell<i32>`
26    = note: required because of the requirements on the impl of `UnwindSafe` for `Rc<RefCell<i32>>`
- error: aborting due to 2 previous errors
- error: aborting due to 2 previous errors
+ error[E0277]: the type `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
+    |
+    |
+ LL | fn assert<T: UnwindSafe + ?Sized>() {}
+    |              ---------- required by this bound in `assert`
+ ...
+ LL |     assert::<Rc<RefCell<i32>>>();
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
+    |
+    = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<Option<&'static Location<'static>>>`
+    = note: required because it appears within the type `Cell<Option<&'static Location<'static>>>`
+    = note: required because it appears within the type `RefCell<i32>`
+    = note: required because of the requirements on the impl of `UnwindSafe` for `Rc<RefCell<i32>>`
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
+ error: aborting due to 3 previous errors
29 
30 For more information about this error, try `rustc --explain E0277`.
30 For more information about this error, try `rustc --explain E0277`.
31 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-2/not-panic-safe-2.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args not-panic/not-panic-safe-2.rs`
error: 1 errors occurred comparing output.
status: exit code: 1
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/not-panic/not-panic-safe-2.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-2" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-2/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0277]: the type `UnsafeCell<i32>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<Rc<RefCell<i32>>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<i32>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<i32>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `Rc<RefCell<i32>>`

error[E0277]: the type `UnsafeCell<isize>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<Rc<RefCell<i32>>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<isize>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<isize>`
   = note: required because it appears within the type `Cell<isize>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `Rc<RefCell<i32>>`

error[E0277]: the type `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<Rc<RefCell<i32>>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<Option<&'static Location<'static>>>`
   = note: required because it appears within the type `Cell<Option<&'static Location<'static>>>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `Rc<RefCell<i32>>`
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.


------------------------------------------


---- [ui] ui/not-panic/not-panic-safe-4.rs stdout ----
diff of stderr:

25    = note: required because it appears within the type `RefCell<i32>`
26    = note: required because of the requirements on the impl of `UnwindSafe` for `&RefCell<i32>`
- error: aborting due to 2 previous errors
- error: aborting due to 2 previous errors
+ error[E0277]: the type `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
+    |
+    |
+ LL | fn assert<T: UnwindSafe + ?Sized>() {}
+    |              ---------- required by this bound in `assert`
+ ...
+ LL |     assert::<&RefCell<i32>>();
+    |     ^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
+    |
+    = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<Option<&'static Location<'static>>>`
+    = note: required because it appears within the type `Cell<Option<&'static Location<'static>>>`
+    = note: required because it appears within the type `RefCell<i32>`
+    = note: required because of the requirements on the impl of `UnwindSafe` for `&RefCell<i32>`
+ error: aborting due to 3 previous errors
29 
30 For more information about this error, try `rustc --explain E0277`.
31 
31 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-4/not-panic-safe-4.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args not-panic/not-panic-safe-4.rs`
error: 1 errors occurred comparing output.
status: exit code: 1
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/not-panic/not-panic-safe-4.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-4" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-4/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0277]: the type `UnsafeCell<i32>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<&RefCell<i32>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<i32>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<i32>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `&RefCell<i32>`

error[E0277]: the type `UnsafeCell<isize>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<&RefCell<i32>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<isize>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<isize>`
   = note: required because it appears within the type `Cell<isize>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `&RefCell<i32>`

error[E0277]: the type `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<&RefCell<i32>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<Option<&'static Location<'static>>>`
   = note: required because it appears within the type `Cell<Option<&'static Location<'static>>>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `&RefCell<i32>`
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.


------------------------------------------


---- [ui] ui/not-panic/not-panic-safe-3.rs stdout ----
diff of stderr:

25    = note: required because it appears within the type `RefCell<i32>`
26    = note: required because of the requirements on the impl of `UnwindSafe` for `Arc<RefCell<i32>>`
- error: aborting due to 2 previous errors
- error: aborting due to 2 previous errors
+ error[E0277]: the type `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
+    |
+    |
+ LL | fn assert<T: UnwindSafe + ?Sized>() {}
+    |              ---------- required by this bound in `assert`
+ ...
+ LL |     assert::<Arc<RefCell<i32>>>();
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
+    |
+    = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<Option<&'static Location<'static>>>`
+    = note: required because it appears within the type `Cell<Option<&'static Location<'static>>>`
+    = note: required because it appears within the type `RefCell<i32>`
+    = note: required because of the requirements on the impl of `UnwindSafe` for `Arc<RefCell<i32>>`
+ error: aborting due to 3 previous errors
29 
30 For more information about this error, try `rustc --explain E0277`.
31 
31 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-3/not-panic-safe-3.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args not-panic/not-panic-safe-3.rs`
error: 1 errors occurred comparing output.
status: exit code: 1
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/not-panic/not-panic-safe-3.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-3" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-3/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0277]: the type `UnsafeCell<i32>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<Arc<RefCell<i32>>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<i32>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<i32>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `Arc<RefCell<i32>>`

error[E0277]: the type `UnsafeCell<isize>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<Arc<RefCell<i32>>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<isize>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<isize>`
   = note: required because it appears within the type `Cell<isize>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `Arc<RefCell<i32>>`

error[E0277]: the type `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<Arc<RefCell<i32>>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<Option<&'static Location<'static>>>`
   = note: required because it appears within the type `Cell<Option<&'static Location<'static>>>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `Arc<RefCell<i32>>`
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.


------------------------------------------


---- [ui] ui/not-panic/not-panic-safe-6.rs stdout ----
diff of stderr:

25    = note: required because it appears within the type `RefCell<i32>`
26    = note: required because of the requirements on the impl of `UnwindSafe` for `*mut RefCell<i32>`
- error: aborting due to 2 previous errors
- error: aborting due to 2 previous errors
+ error[E0277]: the type `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
+    |
+    |
+ LL | fn assert<T: UnwindSafe + ?Sized>() {}
+    |              ---------- required by this bound in `assert`
+ ...
+ LL |     assert::<*mut RefCell<i32>>();
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
+    |
+    = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<Option<&'static Location<'static>>>`
+    = note: required because it appears within the type `Cell<Option<&'static Location<'static>>>`
+    = note: required because it appears within the type `RefCell<i32>`
+    = note: required because of the requirements on the impl of `UnwindSafe` for `*mut RefCell<i32>`
+ error: aborting due to 3 previous errors
29 
30 For more information about this error, try `rustc --explain E0277`.
31 
31 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-6/not-panic-safe-6.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args not-panic/not-panic-safe-6.rs`
error: 1 errors occurred comparing output.
status: exit code: 1
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/not-panic/not-panic-safe-6.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-6" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/not-panic/not-panic-safe-6/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0277]: the type `UnsafeCell<i32>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<*mut RefCell<i32>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<i32>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<i32>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `*mut RefCell<i32>`

error[E0277]: the type `UnsafeCell<isize>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<*mut RefCell<i32>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<isize>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<isize>`
   = note: required because it appears within the type `Cell<isize>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `*mut RefCell<i32>`

error[E0277]: the type `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   |
LL | fn assert<T: UnwindSafe + ?Sized>() {}
   |              ---------- required by this bound in `assert`
...
LL |     assert::<*mut RefCell<i32>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<Option<&'static Location<'static>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `RefCell<i32>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<Option<&'static Location<'static>>>`
   = note: required because it appears within the type `Cell<Option<&'static Location<'static>>>`
   = note: required because it appears within the type `RefCell<i32>`
   = note: required because of the requirements on the impl of `UnwindSafe` for `*mut RefCell<i32>`
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.

---
test result: FAILED. 11365 passed; 4 failed; 93 ignored; 0 measured; 0 filtered out; finished in 134.54s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-9/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "9.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver engine executionengine fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:16:15
```

</details>

estebank commented on library/core/src/cell.rs:
```diff
@@ -793,7 +803,17 @@ impl<T: ?Sized> RefCell<T> {
     pub fn borrow(&self) -> Ref<'_, T> {
-        self.try_borrow().expect("already mutably borrowed")
+        // FIXME: Use `unwrap_or_else` once we support `#[track_caller]` on closures
+        #[cfg(feature = "debug_refcell")]
+        match self.try_borrow() {
+            Ok(b) => return b,
+            Err(e) => {
+                panic!("already mutably borrowed: {:?} at {}", e, self.first_caller.get().unwrap())
+            }
+        }
+
+        #[cfg(not(feature = "debug_refcell"))]
+        return self.try_borrow().expect("already mutably borrowed");
```
Wouldn't it be better if the `cfg` only covered the actual panic line? As it stands there's "hidden" duplication across feature gated state.

estebank commented on library/core/src/cell.rs:
```diff
@@ -823,11 +843,21 @@ impl<T: ?Sized> RefCell<T> {
+            Some(b) => {
+                #[cfg(feature = "debug_refcell")]
+                {
+                    if b.borrow.get() == 1 {
+                        self.first_caller.set(Some(crate::panic::Location::caller()));
```
I was gonna say the following, but the closure would screw the `caller` call, right?

---

No idea if this is more "efficient", but an alternative here would be

`self.first_caller.update(|caller| caller.or(Some(crate::panic::Location::caller())))`

Aaron1011 replied:
Your suggestion won't work - `crate::panic::Location::caller()` will hit the boundary of the `|caller|` closure.

Mark-Simulacrum commented:
Implementation seems reasonable - I'm not sure how much we want to do this, but I could easily see it being quite useful in practice.

r? @m-ou-se for T-libs

Aaron1011 commented:
Ancedotally, I've already used this to fix a bug (https://github.com/ruffle-rs/ruffle/pull/3322) where the `borrow()` and `borrow_mut()` calls were very far apart in the call stack.

m-ou-se commented on library/core/Cargo.toml:
```diff
@@ -25,3 +25,6 @@ rand = "0.7"
 [features]
 # Make panics and failed asserts immediately abort without formatting any message
 panic_immediate_abort = []
+debug_refcell = []
```
Can you add a comment here that briefly explains what it does?

m-ou-se commented on library/core/src/cell.rs:
```diff
@@ -569,12 +569,17 @@ impl<T> Cell<[T]> {
 pub struct RefCell<T: ?Sized> {
     borrow: Cell<BorrowFlag>,
+    #[cfg(feature = "debug_refcell")]
+    first_caller: CallerLoc,
```
I'd just write out the type here, instead of using an alias that's only used once. (Saves a `#[cfg]` too.)

m-ou-se commented on library/core/src/cell.rs:
```diff
@@ -658,7 +663,12 @@ impl<T> RefCell<T> {
+            value: UnsafeCell::new(value),
+            borrow: Cell::new(UNUSED),
+            #[cfg(feature = "debug_refcell")]
+            first_caller: Cell::new(None),
```
The name `first_caller` isn't very obvious to me. How about something like `borrowed_by` or `borrowed_at` or `borrowed_first_at`?

m-ou-se reviewed (commented):
> * The captured location is only displayed when `borrow()` or
>   `borrow_mut()` panics. If a crate calls `try_borrow().unwrap()`
>   or `try_borrow_mut().unwrap()`, this extra information will be lost.

Why not add the location to the `BorrowError` and `BorrowMutError` types instead? Then you don't need to modify the `borrow()` and `borrow_mut()` functions at all, as the `.expect()` will include the information in the panic message. And it'll make `try_borrow().unwrap()` (etc.) also show the information.

Aaron1011 commented:
@m-ou-se I've addressed your comments. I've only modified the `Debug` impls of `BorrowError`/`BorrowMutError` - I've left the `Display` impls alone for now.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2104274346) [(plain)](https://github.com/rust-lang/rust/commit/fbdaf3fb43078fc31a819f2f5de2b5ca2cef75dd/checks/2104274346/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0308]: mismatched types
  --> /checkout/compiler/rustc_data_structures/src/macros.rs:5:32
   |
3  | / macro_rules! static_assert_size {
4  | |     ($ty:ty, $size:expr) => {
5  | |         const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
   | |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 152 elements, found one with 168 elements
7  | | }
   | |_- in this expansion of `rustc_data_structures::static_assert_size!`
   | 
  ::: src/librustdoc/html/render/context.rs:79:1
  ::: src/librustdoc/html/render/context.rs:79:1
   |
79 |   rustc_data_structures::static_assert_size!(Context<'_>, 152);

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```

</details>

Aaron1011 commented:
The assertion failure is due to the fact that I've enabled this by default to make testing easier.

m-ou-se commented on library/core/src/cell.rs:
```diff
@@ -869,7 +911,7 @@ impl<T: ?Sized> RefCell<T> {
     #[inline]
     #[track_caller]
     pub fn borrow_mut(&self) -> RefMut<'_, T> {
-        self.try_borrow_mut().expect("already borrowed")
+        return self.try_borrow_mut().expect("already borrowed");
```
Looks like this line was changed by accident?

m-ou-se reviewed (commented):
This looks good!

r=me with the feature disabled by default.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2168006492) [(plain)](https://github.com/rust-lang/rust/commit/a8ef043216bf2c0de22abe6b7f6f563d4b649514/checks/2168006492/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error[E0308]: mismatched types
   --> library/core/src/cell.rs:820:29
    |
820 |     pub fn borrow(&self) -> Ref<'_, T> {
    |            ------           ^^^^^^^^^^ expected struct `Ref`, found `()`
    |            |
    |            implicitly returns `()` as its body has no tail or `return` expression
821 |         self.try_borrow().expect("already mutably borrowed");
    |                                                             - help: consider removing this semicolon
    |
    = note: expected struct `Ref<'_, T>`

error[E0308]: mismatched types
   --> library/core/src/cell.rs:913:33
    |
    |
913 |     pub fn borrow_mut(&self) -> RefMut<'_, T> {
    |            ----------           ^^^^^^^^^^^^^ expected struct `RefMut`, found `()`
    |            |
    |            implicitly returns `()` as its body has no tail or `return` expression
914 |         self.try_borrow_mut().expect("already borrowed");
    |                                                         - help: consider removing this semicolon
    |
    = note: expected struct `RefMut<'_, T>`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
For more information about this error, try `rustc --explain E0308`.
error: could not compile `core`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:01:34
```

</details>

Aaron1011 commented:
@bors r=m-ou-se

bors commented:
:pushpin: Commit a23273e82c8156974553e43f7d61826651397fbc has been approved by `m-ou-se`

<!-- @bors r=m-ou-se a23273e82c8156974553e43f7d61826651397fbc -->
<!-- homu: {"type":"Approved","sha":"a23273e82c8156974553e43f7d61826651397fbc","approver":"m-ou-se"} -->

bors commented:
:hourglass: Testing commit a23273e82c8156974553e43f7d61826651397fbc with merge 9b6339e4b9747d473270baa42e77e1d2fff39bf4...
<!-- homu: {"type":"BuildStarted","head_sha":"a23273e82c8156974553e43f7d61826651397fbc","merge_sha":"9b6339e4b9747d473270baa42e77e1d2fff39bf4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2173137491)
Approved by: m-ou-se
Pushing 9b6339e4b9747d473270baa42e77e1d2fff39bf4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"m-ou-se","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2173137491"},"merge_sha":"9b6339e4b9747d473270baa42e77e1d2fff39bf4"} -->
